libcloud/dns/drivers/gandi_live.py

```diff
@@ -0,0 +1,347 @@
+
+
+class GandiLiveDNSDriver(BaseGandiLiveDriver, DNSDriver):
+    """
+    API reference can be found at:
+
+    https://doc.livedns.gandi.net/
+
+    Please note that the Libcloud paradigm of one zone per domain does not
+    match exactly with Gandi LiveDNS.  For Gandi, a "zone" can apply to
+    multiple domains.  This driver behaves as if the domain is a zone, but be
+    warned that modifying a domain means modifying the zone.  Iif you have a
+    zone associated with mutiple domains, all of those domains will be
+    modified as well.
+    """
+
+    type = Provider.GANDI
+    name = 'Gandi LiveDNS'
+    website = 'http://www.gandi.net/domain'
+
+    connectionCls = GandiLiveDNSConnection
+
+    # also supports CAA, CDS
+    RECORD_TYPE_MAP = {
+        RecordType.A: 'A',
+        RecordType.AAAA: 'AAAA',
+        RecordType.ALIAS: 'ALIAS',
+        RecordType.CNAME: 'CNAME',
+        RecordType.DNAME: 'DNAME',
+        RecordType.DS: 'DS',
+        RecordType.KEY: 'KEY',
+        RecordType.LOC: 'LOC',
+        RecordType.MX: 'MX',
+        RecordType.NS: 'NS',
+        RecordType.PTR: 'PTR',
+        RecordType.SPF: 'SPF',
+        RecordType.SRV: 'SRV',
+        RecordType.SSHFP: 'SSHFP',
+        RecordType.TLSA: 'TLSA',
+        RecordType.TXT: 'TXT',
+        RecordType.WKS: 'WKS',
+    }
+
+    def _to_zone(self, zone):
+        extra = {}
+        if 'zone_uuid' in zone:
+            extra = {
+                'zone_uuid': zone['zone_uuid']
+            }
+        return Zone(
+            id=str(zone['fqdn']),
+            domain=zone['fqdn'],
+            type='master',
+            ttl=0,
+            driver=self,
+            extra=extra,
+        )
+
+    def _to_zones(self, zones):
+        ret = []
+        for z in zones:
+            ret.append(self._to_zone(z))
+        return ret
+
+    def list_zones(self):
+        zones = self.connection.request(action='%s/domains' % API_BASE,
+                                        method='GET')
+        return self._to_zones(zones.object)
+
+    def get_zone(self, zone_id):
+        action = '%s/domains/%s' % (API_BASE, zone_id)
+        try:
+            zone = self.connection.request(action=action, method='GET')
+        except ResourceNotFoundError:
+            raise ZoneDoesNotExistError(value='',
+                                        driver=self.connection.driver,
+                                        zone_id=zone_id)
+        return self._to_zone(zone.object)
+
+    """
+    :param extra: (optional) Extra attributes ('name'); if not provided, name
+                             is based on domain.
+    """
+    def create_zone(self, domain, type='master', ttl=None, extra=None):
+        if extra and 'name' in extra:
+            zone_name = extra['name']
+        else:
+            zone_name = '%s zone' % domain
+        raw_zone_data = {
+            'name': zone_name,
+        }
+        post_zone_data = json.dumps(raw_zone_data)
```

Comment:
This encoding should happens inside the connection class - please add ``encode_data`` to the connection class and remove data encoding inside the driver methods.

For details and context, please see https://github.com/apache/libcloud/pull/1315#issuecomment-512226526

Reply:
Ah, didn't realize that was a thing.  Fixed.